<div class="alert alert-block alert-success">
<b>Comentario general del revisor</b> <a class="tocSkip"></a><br />
Status del proyecto: <b>Aprobado</b>
</div>

¡Hola!<br />
Soy **Francisco Cortés**, estoy contento de revisar tu proyecto y ser parte de tu proceso de aprendizaje.
A lo largo del texto, haré algunas observaciones sobre mejoras en el código y también haré comentarios sobre tus percepciones sobre el tema. Si existe algún error en el código, no te preocupes, estoy aquí para ayudarte a mejorarlo, en la primera iteración te lo señalaré para que tengas la oportunidad de corregirlo, pero si aún no encuentras una solución para esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a retroalimentación y discusiones sobre el tema.<br />
Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o borres**.
Revisaré cuidadosamente tu código para comprobar que se han cumplido con los requisitos y te proporcionaré mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si la ejecución fue perfecta.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si existe alguna recomendación para que tu código mejore.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si existen correcciones necesarias para cumplir con los requisitos. El trabajo no puede ser aceptado si hay alguna caja roja.
</div>

Puedes responderme de la siguiente manera:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante.</b> <a class="tocSkip"></a>
</div>


# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df
df=pd.read_csv('/datasets/music_project_en.csv')

Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Correcto!<br/>

Buena forma de leer los datos y hacer una revisión inicial

</div>


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.Lo unico que me hace ruido y necesito revisar antes de continuar, parece que Day; revisando me encontre con que Day sis e puede usar, pero time no, por que es el nombre de un módulo de Python, por lo que es una mala practica, y puede llegar a causar conflictos, conforme avance en el proyecto detecte que se me habia pasado la columna de userId, que es preferente manejarlo como palabras separadas por guio bajo, para seguir la nomenglatura y reglas.(se reviso busqueda)


### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

1. Tomando una muestras de 10 registros procedimos a realizar obtener algunas conjeturas de lo que estamos encontrando en los registros de usuarios con informacion sobre la musica que les gusta escuchar, se encontro que por columnas:
 - Contamos con 65079 registros, se cuenta con 7 columnas son de tipo object
 - Cada registro contiene id de usuario/a, titulo de la cancion, nombre del artista, el genero, la ciudad donde el usuario/a la esta escuchando, la hr en que se reprodujo la cancion y el día.
 
 
2. Creo que si hay suficiente información, aunque me gustaria contar con una columna que proporcinara la edad para saber el rango de edad que escucha ese genero o determinado artista, ya que si las tomamos encuenta de que año es la cancion son un sin fin de canciones.


3. Si existe algunos problemas tanto en algunas columnas con respecto a los nombres, como a los datos, encontre varios valores ausentes para canciones, artistas y genero y tambien encontre valores duplicados, de los datos incorrectos con los pocos datos que he visto no encontrado, pero si deben de tener.





<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Bien hecho!<br/>

Las observaciones que haces son buenas

</div>


[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle en los encabezados poniendo todo en minúsculas
new_columns = []
for col in df.columns:
    col=col.lower()
    new_columns.append(col)
df.columns = new_columns
print(df.columns)


Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [7]:
# Bucle en los encabezados eliminando los espacios
new_columns = []
for col in df.columns:
    col=col.strip()
    new_columns.append(col)
df.columns = new_columns
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambiar el nombre de la columna "userid"
new_columns = []
for col in df.columns:
    if col =='userid':
        col='user_id'
    new_columns.append(col)
df.columns = new_columns
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprobar el resultado: la lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Correcto!<br/>
Buena manera de corregir y reemplazar los nombres de las columnas
</div>


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [10]:
# Calcular el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
#columnas a tratar
colum_cambiar =['track', 'artist','genre']
for col in df.columns:
    if col in colum_cambiar:
        df[col]=df[col].fillna('unknown', inplace =False)
print(df.isna().sum())


user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [12]:
# Contar valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Correcto!<br/>
Buena manera de reemplazar los valores ausentes en el dataframe
</div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [13]:
# Contar duplicados explícitos
print(df.duplicated().sum())

3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Eliminar duplicados explícitos
df.drop_duplicates(inplace=True)

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprobar de nuevo si hay duplicados
print(df.duplicated().sum())


0


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Bien hecho!<br/>
Buena manera de eliminar los duplicados explicitos
</div>

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [17]:
# Inspeccionar los nombres de géneros únicos
print(sorted(df['genre'].unique()))
print()
print("número de genre unicos ------>",df['genre'].nunique())

#print(df['genre'])


['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [18]:

#print(df['genre'])
print("contador para hiphop")
print((df['genre']=='hiphop').sum())
print()
print("contador para hip")
print((df['genre']=='hip').sum())
print()
print("contador para hip-hop")
print((df['genre']=='hip-hop').sum())
print()


contador para hiphop
79

contador para hip
2975

contador para hip-hop
1



In [19]:
# Función para reemplazar duplicados implícitos
#wrong_genres=: esta es una lista que contiene todos los valores que necesitas reemplazar.
#correct_genre=: este es un string que vas a utilizar como reemplazo.

def replace_wrong_genres(df, wrong_genres, correct_genre):
    for index, row in df.iterrows():
        if row['genre'] in wrong_genres:
            df.at[index, 'genre'] = correct_genre
        
    return df


Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [20]:
# Eliminar duplicados implícitos
lista =['hip','hop','hip-hop']

   
datos_remplazados = replace_wrong_genres(df, lista, 'hiphop')
print(datos_remplazados)

        user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfi

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [21]:
# Comprobación de duplicados implícitos

#print(df['genre'])
print("contador para hiphop")
print((df['genre']=='hiphop').sum())
print()
print("contador para hip")
print((df['genre']=='hip').sum())
print()
print("contador para hip-hop")
print((df['genre']=='hip-hop').sum())
print()


contador para hiphop
3056

contador para hip
0

contador para hip-hop
0



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Correcto, se eliminaron los duplicados explicitos de manera adecuada y los duplicados implicitos se reemplazaron por un valor estandarizado, los duplicados implicitos por lo general son más difíciles de encontrar debido a que se requiere contexto/conocimiento de los datos que se estan analizando, bien hecho
</div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`

Y lo que veo que que hay muchos valores diferentes de genero en esa columna, por lo que no se si se esta llevando bien toda la limpieza de datos, creo que deberia proporcionar una lista de todas los generos de musica que consideran encontrar y entonces, partiendo de eso haceer un barrido de todo y partir de que si no se encuentra en ese catalogo proceder a cambiar de genero o eliminar esos registro, ya que solo generan confunción.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [23]:
print("usuarios/as de Springfield")
df_sprnnfie = df.query("city == 'Springfield'")['day'].count()
print(df_sprnnfie)

    
#df_sprnnfie = df.query("city == 'Springfield' and 'day=='Monday'")['day'].count()
df_springfield_lune = df.query("city == 'Springfield' and day == 'Monday'")['day'].count()
print('Lunes--->',df_springfield_lune)

df_springfield_mier = df.query("city == 'Springfield' and day == 'Wednesday'")['day'].count()
print('miercoles--->',df_springfield_mier)

df_springfield_fri = df.query("city == 'Springfield' and day == 'Friday'")['day'].count()
print('viernes--->',df_springfield_fri)
    
print()
print("usuarios/as de Shelbyville")

df_shelby = df.query("city == 'Shelbyville'")['day'].count()
print(df_shelby)

    
#df_sprnnfie = df.query("city == 'Springfield' and 'day=='Monday'")['day'].count()
df_shelby_lune = df.query("city == 'Shelbyville' and day == 'Monday'")['day'].count()
print('Lunes--->',df_shelby_lune)

df_shelby_mier = df.query("city == 'Shelbyville' and day == 'Wednesday'")['day'].count()
print('miercoles--->',df_shelby_mier)

df_shelby_fri = df.query("city == 'Shelbyville' and day == 'Friday'")['day'].count()
print('viernes--->',df_shelby_fri)

usuarios/as de Springfield
42741
Lunes---> 15740
miercoles---> 11056
viernes---> 15945

usuarios/as de Shelbyville
18512
Lunes---> 5614
miercoles---> 7003
viernes---> 5895


In [24]:
df_springfield_grouped = (
    df.query("city == 'Springfield'")
    .groupby('day')['day']
    .count()
    .reindex(['Monday', 'Wednesday', 'Friday']) 
)

print(df_springfield_grouped)

day
Monday       15740
Wednesday    11056
Friday       15945
Name: day, dtype: int64


In [25]:
df_Shelbyville_grouped = (
    df.query("city == 'Shelbyville'")
    .groupby('day')['day']
    .count()
    .reindex(['Monday', 'Wednesday', 'Friday']) 
)

print(df_Shelbyville_grouped)

day
Monday       5614
Wednesday    7003
Friday       5895
Name: day, dtype: int64


In [26]:
# Contar las canciones reproducidas en cada ciudad

print("usuarios/as de Springfield")
df_sprnnfie = df.query("city == 'Springfield'")['day'].count()
print(df_sprnnfie)

print("usuarios/as de Shelbyville")

df_shelby = df.query("city == 'Shelbyville'")['day'].count()
print(df_shelby)

usuarios/as de Springfield
42741
usuarios/as de Shelbyville
18512


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Correcto!<br/>
Buena manera de agrupar los datos y de mostrar los resultados, los caules son los esperados en la última parte del proyecto, eso me indica que tu curiosidad hace que vayas por buen camino y te adelantas a lo que se espera en el proyecto, bien hecho!
</div>

`Comenta tus observaciones aquí`

Bueno de acuerdo a los querys aplicados, en Springfield los usuario escuchan mas musica (42741) a diferencia de Shelbyville (18512)


Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [27]:
# Calcular las canciones reproducidas en cada uno de los tres días

print("usuarios/as de Springfield")

df_springfield_lune = df.query("city == 'Springfield' and day == 'Monday'")['day'].count()
print('Lunes--->',df_springfield_lune)

df_springfield_mier = df.query("city == 'Springfield' and day == 'Wednesday'")['day'].count()
print('miercoles--->',df_springfield_mier)

df_springfield_fri = df.query("city == 'Springfield' and day == 'Friday'")['day'].count()
print('viernes--->',df_springfield_fri)
    
print()    
print("usuarios/as de Shelbyville")

df_shelby_lune = df.query("city == 'Shelbyville' and day == 'Monday'")['day'].count()
print('Lunes--->',df_shelby_lune)

df_shelby_mier = df.query("city == 'Shelbyville' and day == 'Wednesday'")['day'].count()
print('miercoles--->',df_shelby_mier)

df_shelby_fri = df.query("city == 'Shelbyville' and day == 'Friday'")['day'].count()
print('viernes--->',df_shelby_fri)

usuarios/as de Springfield
Lunes---> 15740
miercoles---> 11056
viernes---> 15945

usuarios/as de Shelbyville
Lunes---> 5614
miercoles---> 7003
viernes---> 5895


`Comenta tus observaciones aquí`

Como ya se habia mencionado Springfield escuchan mas canciones, pero independientemente de eso se observa que en ambas ciudades, varian los dias que escuchan mas,por ejemplo Springfield, escuchan mas los Viernes, seguido por lunes y el miercoles deben de estar con mas precion, ya que disminuye, a diferencia de Shelbyville que ellos escuchan mas musica el día miercoles, despues es viernes y por ultimo lunes, supongo que van iniciando semana y no tienen mucho tiempo para escuchar musica por que deben de tener pendientes.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [28]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

def number_tracks(df, day, city):
    # Filtrar por día
    df_filtered = df[df['day'] == day]
    
    # Filtrar por ciudad sobre el resultado anterior
    df_filtered = df_filtered[df_filtered['city'] == city]
    
    # Contar el número de valores en la columna 'user_id'
    total_tracks = df_filtered['user_id'].count()
    
    return total_tracks


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [29]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks(df,'Monday','Springfield')

15740

In [30]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks(df,'Monday','Shelbyville')

5614

In [31]:
# El número de canciones reproducidas en Springfield el miércoles
number_tracks(df,'Wednesday','Springfield')

11056

In [32]:
# El número de canciones reproducidas en Shelbyville el miércoles
number_tracks(df,'Wednesday','Shelbyville')


7003

In [33]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks(df,'Friday','Springfield')


15945

In [34]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks(df,'Friday','Shelbyville')


5895

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Correcto!<br/>
La función que has creado agrupa adecuadamente los datos, bien hecho!
</div>

**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`

La hipótesis  es correcta despues de todo el analisis que se a realizado, no siguen ningun patron, ambas ciudades son diferentes, tanto en los dias que escuchan musica, ya que como se meciona antes son diferentes, deberiamos contar con mas datos, o saber y tienen días festivos, algun puente, cosas que sean impactantes para que sea el motivo de que se escucha mas unos dias que otros.

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`

No se puede generalizar y tratar de encontrar un patron para que se siga, ambos grupos de población son diferentes, incluso sus gustos de musica


# Comentario general del revisor
<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />


    
Has realizado un buen trabajo, me doy cuenta de que has aplicado los conocimientos que has adquirido durante el curso, las conclusiones a las que llegaste son buenas y los procedimientos realizados son correctos, la hipotesis se puede decir que es cierta en este caso limitado de datos, si tuvieramos más datos acerca de las poblaciones tendriamos oportunidad de llegar a conclusiones más certeras.<br/>

Continúa con el buen trabajo y mucho éxito en el siguiente Sprint!
</div>


### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)